drivers: i3c: add synopsys designware i3c driver

XenuIsWatching opened the pull request:
This implements the initial driver for the synopsys designware i3c IP.

aspeeddylan commented:
Is it possible to add platform-specific infrastructure, as has been done in the Linux kernel?

https://github.com/torvalds/linux/commit/d782188cbb05a196e46a4838484f020ceeb889ec

XenuIsWatching commented:
> Is it possible to add platform-specific infrastructure, as has been done in the Linux kernel?
> 
> [torvalds/linux@d782188](https://github.com/torvalds/linux/commit/d782188cbb05a196e46a4838484f020ceeb889ec)

I don't know how you would want that within Zephyr. That appears to be Linux OS specific stuff....

XenuIsWatching commented:
An update on this... I can no longer work on this as I have had a "change in direction" internally.
This PR is still up if anyone wants to 'take over' and debug and test (which I never got around to doing).

XenuIsWatching commented:
I had to abandon this due to a change of direction... This will remain open though. Talking privately with some people, this PR may be taken over by those... as this appears to be the most complete driver off all the synopsis I3C DW PRs that I see here.

All that was remaining was to test and debug this driver

XenuIsWatching commented:
This appears to be the most complete of all the Synopsys I3C PRs out there (although I may be biased towards myself 😛 ) As this does support target mode, IBI TIRs, HDR-DDR, HDR-TSP, HDR-TSL, proper attachment and detachment of devices, and short CCCs. I have heard from some that I have planned to try this out.... although no timeline is known.

XenuIsWatching committed:
drivers: i3c: add snps_designware_i3c to i3c shell

add the synopsys designware i3c to the i3c shell.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

decsny commented on dts/bindings/i3c/snps,designware-i3c.yaml:
```diff
@@ -0,0 +1,26 @@
+
+  od-thigh-max-ns:
+    type: int
+    default: 41
```
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

XenuIsWatching replied:
This comes from the I3C Specification.... but I can add a note there

decsny replied:
yes, usually defaults do have some justification behind them, that's why we need to put it in the description

XenuIsWatching commented:
This is dependent on a change in https://github.com/zephyrproject-rtos/zephyr/pull/71170 as to why the CI is failing

gmarull commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2346 @@
+	} else if (hcnt > SCL_I3C_TIMING_CNT_MAX) {
+		hcnt = SCL_I3C_TIMING_CNT_MAX;
+	}
+	lcnt = DIV_ROUND_UP(core_rate, I3C_BUS_TYP_I3C_SCL_RATE) - hcnt;
```
```suggestion
	lcnt = DIV_ROUND_UP(core_rate, data->common.ctrl_config.scl.i3c) - hcnt;
```

XenuIsWatching replied:
How the IP wants to be defined is that it uses the MXDS rate (8, 6, 4, 2 MHz....) and just 'assumes' that everything can get up to 12.5MHz.... but I understand your point with how 'sometimes' a user will want to have a configured rate there

gmarull commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2346 @@
+	static struct i3c_i2c_device_desc dw_i3c_i2c_device_array_##n[] =                          \
+		I3C_I2C_DEVICE_ARRAY_DT_INST(n);                                                   \
+	static struct dw_i3c_data dw_i3c_data_##n = {                                              \
+		.common.ctrl_config.scl.i3c = DT_INST_PROP_OR(n, i3c_scl_hz, 0),                   \
```
```suggestion
		.common.ctrl_config.scl.i3c = DT_INST_PROP_OR(n, i3c_scl_hz, I3C_BUS_TYP_I3C_SCL_RATE),                   \
```

gmarull commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2346 @@
+
+	/* I3C_OD */
+	hcnt = DIV_ROUND_UP(config->od_thigh_max_ns * core_rate, I3C_PERIOD_NS) - 1;
+	if (hcnt < SCL_I3C_TIMING_CNT_MIN) {
+		hcnt = SCL_I3C_TIMING_CNT_MIN;
+	} else if (hcnt > SCL_I3C_TIMING_CNT_MAX) {
+		hcnt = SCL_I3C_TIMING_CNT_MAX;
+	}
```
```suggestion
	hcnt = CLAMP(hcnt, SCL_I3C_TIMING_CNT_MIN, SCL_I3C_TIMING_CNT_MAX);
```

... and others

gmarull commented on dts/bindings/i3c/snps,designware-i3c.yaml:
```diff
@@ -0,0 +1,32 @@
+description: >
+    This is a binding for the Synopsys I3C interface
+
+compatible: "snps,designware-i3c"
```
shouldn't we maintain compatiblity with Linux?, ie

```suggestion
compatible: "snps,dw-i3c-master-1.00a"
```

XenuIsWatching replied:
I'm not sure if that is necessary, I'm going on constancy here. like how `snps,designware-i2c` is already defined

XenuIsWatching replied:
yeah..., but this is not the 1.00a version of the IP this driver is supporting here. This is the 1.01a which does include some features that are within this driver.

gmarull replied:
so then we need to specify IP version in the compatible, we, for example, have 1.00a. Driver should be able to support both, based on the DT compatible.

XenuIsWatching replied:
looking again through the databook, it looks like both should be supported anyways here

gmarull replied:
I think this is still valid, we have a rule that says we should align with upstream Linux when possible.

XenuIsWatching replied:
but then there is a rule to not introduce 'offensive terms' with the word `master`. :shrug: 
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

I'm still not convinced that linux made the right choice here with the naming of it especially since this driver doesn't only support the "controller" implemention of the IP. I'll poke around on the linux mailing lists to see what response I get and update the PR depending on what I get.

gmarull commented on dts/bindings/i3c/snps,designware-i3c.yaml:
```diff
@@ -0,0 +1,32 @@
+include: i3c-controller.yaml
+
+properties:
+  input-clock-frequency:
+    required: true
+    type: int
+    description: clk frequency.
```
this is a misuse of DT, we should just take a `clocks` property and obtain its rate.

XenuIsWatching replied:
I really wish that was possible, but using the clock api is rather vendor specific and how the clock is obtain would be implementation defined for this IP

gmarull replied:
I agree, clock control API is not _ready_ for mainstream use. I wonder if we should introduce a wrapper pattern like in CAN MCAN, where clock info is part of the wrapper ops. Note that for Nordic we'll need some minimal wrapper code (IP enablement, and IRQ clearing procedure).

XenuIsWatching replied:
I took a look at the CAN MCAN driver... but I don't see what you are referring to with the wrapper pattern..

As for obtaining the input `sys_clk` rate to the ip... I don't see the current clock_control api working from a 'vendor-agnostic' ip such as synopsys designment.... but I do see the RFC clock management API https://github.com/zephyrproject-rtos/zephyr/pull/72102 working for something like this when that comes

gmarull replied:
Here we should be using `clocks = <&clock_controller ...>;`. I know that Zephyr doesn't have a proper clock control API making DT link smoother, but we should still write proper DT and add the necessary glue code in C in the meantime. For example, in the case of Nordic, the clocks are described like this, so we'd be forced to hack our DT to use `input-clock-frequency` when we already have a properly defined clock controller. In some cases, the frequency may not even be computable at compile time.

XenuIsWatching replied:
this is now completed... 'resolve the conversation' if this is what you were looking for?

gmarull commented on tests/drivers/build_all/i3c/boards/qemu_cortex_m3.overlay:
```diff
@@ -16,4 +16,16 @@
+		interrupts = <5 1>;
+		input-clock-frequency = <200000000>;
+		i3c-scl-hz = <12500000>;
+		i2c-scl-hz = <400000>;
```
where is this prop used in the driver?

XenuIsWatching replied:
it's not...this IP is rather stringent with the specification and has seperate registers for it's FM+ and FM I2C rates

gmarull replied:
if not used/supported, it should be removed.

gmarull commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2346 @@
+	struct dw_i3c_data *data = dev->data;
+	int ret = 0;
+
+	if (config == NULL) {
+		ret = -EINVAL;
+		goto out_configure;
+	}
```
redundant?

XenuIsWatching replied:
I don't see where it's redundant

gmarull replied:
Sounds like passing NULL would be a programming error, not something to be handled

XenuIsWatching replied:
yeah.... I could change to a `__ASSERT_NO_MSG`

XenuIsWatching replied:
i just removed it instead...

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2335 @@
+		xfer->ncmds = payload->targets.num_targets;
+		for (i = 0; i < payload->targets.num_targets; i++) {
+			cmd = &xfer->cmds[i];
+			pos = get_i3c_addr_pos(dev, payload->targets.payloads[i].addr);
```
There was an issue found here where if the `addr` is a static addr in the case of SETDASA, then it won't return the correct `pos` as it does the lookup based on the dynamic addr. This will need to be fixed when I can get some time

gmarull replied:
Spotted this while testing the driver on our side (currently on hold due to some issues we're investigating)

XenuIsWatching replied:
> Spotted this while testing the driver on our side (currently on hold due to some issues we're investigating)

Ah, yeah :upside_down_face: . Was there anything else that you spotted?

gmarull replied:
Not really, but due to some other HW issues, I haven't been able to test other advanced features.

XenuIsWatching replied:
ah... we I pushed up a fix for this since then

XenuIsWatching commented:
Fixes that were just committed:

Fixed issue with dat pos look up with static address for the ccc setdasa 
Fixed reattach clearing ibi bits in the dat. Now only modifies the dynamic address in the table rather than clearing the other bits.

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2349 @@
+	hcnt = DIV_ROUND_UP(config->od_thigh_max_ns * core_rate, I3C_PERIOD_NS) - 1;
+	hcnt = CLAMP(hcnt, SCL_I3C_TIMING_CNT_MIN, SCL_I3C_TIMING_CNT_MAX);
+
+	lcnt = DIV_ROUND_UP(config->od_tlow_min_ns * core_rate, I3C_PERIOD_NS);
```
Will it not overflow when:
1. core_rate is 100MHz
2. config->od_tlow_min_ns = 200

XenuIsWatching replied:
this was something we caught internally a while ago, I just haven't push a new commit fixing that issue... but the solution was to just cast `core_rate` to a `uint64_t`

Shreehari-AlifSemi reviewed (commented):
Kindly check the comments

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2349 @@
+		}
+		/* Read Requested when the CMDQ is empty*/
+		if (status & INTR_READ_REQ_RECV_STAT) {
+			if (target_cb != NULL && target_cb->write_received_cb != NULL) {
```
Should it not be "target_cb->read_requested_cb != NULL" ?

XenuIsWatching replied:
no, you just check if target_cb is not NULL before you even look at target_cb->read_requested_cb, otherwise you are dereferencing a NULL will do a hard crash if you do

Shreehari-AlifSemi replied:
/* Read Requested when the CMDQ is empty*/
        if (status & INTR_READ_REQ_RECV_STAT) {
            if (target_cb != NULL && **target_cb->write_received_cb** != NULL) {
                /* TODO: how should the first byte be handled? */
                **target_cb->read_requested_cb**(data->target_config, NULL);
            }
            sys_write32(INTR_READ_REQ_RECV_STAT, config->regs + INTR_STATUS);
        }

Here, validity of "write_received_cb" is checked but "read_requested_cb" is processed, which is wrong. I was pointing this.

XenuIsWatching replied:
ok, I see now, yes that is correct. thanks for pointing that out

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2349 @@
+	newdevs &= ~olddevs;
+
+	for (pos = find_lsb_set(newdevs); pos <= find_msb_set(newdevs); pos++) {
+		if (newdevs & BIT(pos)) {
```
No? It's been so long since I wrote this.... the PR is over a year old now, I did a quick skim through the databook again.

What makes you think it should be `pos-1`?

Shreehari-AlifSemi replied:
Please see the explanation of "find_lsb_set" function:
https://elixir.bootlin.com/zephyr/v4.0.0/source/include/zephyr/arch/common/ffs.h#L41

It says, "This routine finds the first bit set starting from the least significant bit
 * in the argument passed in and returns the index of that bit.  **Bits are
 * numbered starting at 1 from the least significant bit.  A return value of
 * zero indicates that the value passed is zero."**

XenuIsWatching replied:
ah, i see. done

Shreehari-AlifSemi commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2349 @@
+
+	/* retrieve dynamic address assigned */
+	tmp = sys_read32(config->regs + DEV_ADDR_TABLE_LOC(data->datstartaddr, pos));
+	dyn_addr = DEV_ADDR_TABLE_DYNAMIC_ADDR(tmp);
```
It will not work. The actual 7-bit dyn_addr is derived from Bit 16 to Bit 22

XenuIsWatching replied:
ah, it probably still works... but your right, it should be from 22 to 16.

Shreehari-AlifSemi commented:
Please resolve the merge conflict

syedsk commented:
The write_requested_cb callbacks are not implemented for the target mode?

XenuIsWatching commented:
> The write_requested_cb callbacks are not implemented for the target mode?

There is no signal for this, therefore it is unused

syedsk commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2351 @@
+			SLV_SIR_DATA_BYTE2((request->payload_len > 3) ? request->payload[3] : 0);
+		tir_data |=
+			SLV_SIR_DATA_BYTE3((request->payload_len > 4) ? request->payload[4] : 0);
+		sys_write32(config->regs + SLV_SIR_DATA, tir_data);
```
Must be sys_write32(tir_data, config->regs + SLV_SIR_DATA);

XenuIsWatching replied:
done

XenuIsWatching commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2351 @@
+
+	/* kick off the ibi tir request */
+	slv_intr_req |= SLV_INTR_REQ_SIR;
+	sys_write32(config->regs + SLV_INTR_REQ, slv_intr_req);
```
done

XenuIsWatching commented:
There was another issue found where for i2c transactions it will transmit the I3C broadcast header (0x7E). This is now fixed.

pdgendt commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2378 @@
+	return 0;
+}
+
+static const struct i3c_driver_api dw_i3c_api = {
```
Wrap the API using the `DEVICE_API` macro.

```suggestion
static DEVICE_API(i3c, dw_i3c_api) = {
```

XenuIsWatching replied:
done

pdgendt commented on include/zephyr/drivers/i3c.h:
```diff
@@ -1213,6 +1213,21 @@ struct i3c_device_desc *i3c_dev_list_find(const struct i3c_dev_list *dev_list,
+ * @return Pointer to the I3C target device descriptor, or
+ *         `NULL` if none is found.
+ */
+struct i3c_device_desc *i3c_dev_list_i3c_static_addr_find(const struct device *dev,
+						   uint8_t addr);
```
```suggestion
struct i3c_device_desc *i3c_dev_list_i3c_static_addr_find(const struct device *dev,
							  uint8_t addr);
```

XenuIsWatching replied:
done

pdgendt commented on drivers/i3c/i3c_common.c:
```diff
@@ -216,6 +216,24 @@ struct i3c_device_desc *i3c_dev_list_i3c_addr_find(const struct device *dev,
 	return ret;
 }
 
+struct i3c_device_desc *i3c_dev_list_i3c_static_addr_find(const struct device *dev,
+						   uint8_t addr)
```
```suggestion
struct i3c_device_desc *i3c_dev_list_i3c_static_addr_find(const struct device *dev,
							  uint8_t addr)
```

XenuIsWatching replied:
double post?

pdgendt replied:
No, both header and implementation.

pdgendt replied:
Comment still pending here.

XenuIsWatching replied:
as I was making changes per @dcpleung request, I just went ahead and did this

pdgendt commented on drivers/i3c/i3c_dw.c:
```diff
@@ -0,0 +1,2378 @@
+	xfer->ret = ret;
+
+	if (ret < 0) {
+		sys_write32(RESET_CTRL_RX_FIFO | RESET_CTRL_TX_FIFO | RESET_CTRL_RESP_QUEUE |
+				    RESET_CTRL_CMD_QUEUE,
```
```suggestion
		sys_write32(RESET_CTRL_RX_FIFO | RESET_CTRL_TX_FIFO | RESET_CTRL_RESP_QUEUE |
			    RESET_CTRL_CMD_QUEUE,
```

XenuIsWatching replied:
done

pdgendt reviewed (commented):
Some indenting nits

XenuIsWatching commented:
@Shreehari-AlifSemi It's more helpful to give comments about issues you've found rather than 'alif' as an "entire company" decided to go off and fork the entire zephyr tree and cherry-pick patches like this one and then modify stuff on your own tree.

Shreehari-AlifSemi commented:
Hello @Ryan ***@***.***>,

I have provided comments on your changes in the PR. I recently backported this driver to v3.6 to meet my company's work targets, aiming to stay as close to the open-source version as possible.
As I've been working on this driver for a long time, I also have my own complete implementation. If you hadn't pushed this base code, I would have proceeded with my version.
I've credited your work as the foundation for my implementation. Please let me know if you have any concerns or objections regarding the use of your base code.
If necessary, I'm happy to proceed with my own version instead.

Regards,
Shreehari H K

From: Ryan McClelland ***@***.***>
Sent: 28 January 2025 11:31
To: zephyrproject-rtos/zephyr ***@***.***>
Cc: Shreehari HK ***@***.***>; Mention ***@***.***>
Subject: Re: [zephyrproject-rtos/zephyr] drivers: i3c: add synopsys designware i3c driver (PR #64528)


@Shreehari-AlifSemi<https://github.com/Shreehari-AlifSemi> It's more helpful to give comments about issues you've found rather than 'alif' as an "entire company" decided to go off and fork the entire zephyr tree and cherry-pick patches like this one and then modify stuff on your own tree.

-
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/64528#issuecomment-2617959689>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BJ4I6OF2M5J5F64NWFSW2B32M4MKNAVCNFSM6AAAAAA6UIMNQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJXHE2TSNRYHE>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

Shreehari-AlifSemi commented:
> @Shreehari-AlifSemi It's more helpful to give comments about issues you've found rather than modify stuff on your own tree. Being more communicative helps everyone here. I don't think you want an upstream driver differing from what you've got in that company fork.
> 
> I went ahead and looked through all the good changes in `i3c_dw.c` and brought in most of them along with a few changes here and there... but one thing I don't understand is the usage of `SLV_INTR_REQ_IBI_STS`, where do you see that defined versus using `SLV_IBI_RESP`

I appreciate that you found our code useful! However, I’d like to kindly request that appropriate credit be given for the contributions. It would also be great if you could include my company’s license header in the file’s header to acknowledge the origin of the code, ensuring alignment with proper attribution practices.

XenuIsWatching commented:
> I appreciate that you found our code useful! However, I’d like to kindly request that appropriate credit be given for the contributions. It would also be great if you could include my company’s license header in the file’s header to acknowledge the origin of the code, ensuring alignment with proper attribution practices.

@Shreehari-AlifSemi Yes, I can. Thanks for testing out the target feature btw. Can you please respond to my earlier question tho.

XenuIsWatching commented:
> I have provided comments on your changes in the PR. I recently backported this driver to v3.6 to meet my company's work targets, aiming to stay as close to the open-source version as possible.
As I've been working on this driver for a long time, I also have my own complete implementation. If you hadn't pushed this base code, I would have proceeded with my version.
I've credited your work as the foundation for my implementation. Please let me know if you have any concerns or objections regarding the use of your base code.
If necessary, I'm happy to proceed with my own version instead.

Thank you for the comments and for crediting my work as the foundation for your implementation. I appreciate your transparency in explaining how your team has used and adapted the driver.

That said, I think collaboration would have been more effective if contributions and modifications were communicated earlier. Forking the Zephyr tree and modifying patches independently can lead to fragmentation, making it harder to align with the mainline project and ensure consistency across versions.

If there are specific issues or improvements from your implementation that would benefit the community version, I’d encourage you to share them directly as future contributions. I’m happy to discuss any concerns or questions you have about the base code.

In fact internally, just last week someone was talking to me about to handle that vendor version and how to best handle it with all the other modules floating around.

Shreehari-AlifSemi commented:
> @Shreehari-AlifSemi I appreciate your interest in our code and the effort you put into testing the target feature! Regarding your request for attribution, most of the changes you referenced had incompatibility issues with handoff support and couldn’t be integrated. **As a result, only about 10 code lines were retained, which is quite minor and typically wouldn’t warrant adding a new license header.** Copyright is typically reserved for substantial contributions

I disagree to this.

HemanthKumar-AlifSemi commented:
All, 

Why is this pull request taking a long time to review and merge?

xodus7 reviewed (approved):
We've been using this driver in our project for a while now.

ycsin reviewed (dismissed):
Oh wow, 2023?😅

Code looks clean to me

dcpleung commented:
Could you move the changes under `tests/` from first commit to the last commit?

XenuIsWatching committed:
drivers: i3c: add synopsys designware i3c driver

Add synopsys designware i3c driver

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
tests: drivers: build_all: add build for i3c_dw

The Synopsys I3C has no in-tree board with it to built with. This
adds a test build-only for the cadence i3c peripheral attaching
with the qemu_cortex_m3 board as that appears to be generic
enough.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching commented:
> Could you move the changes under `tests/` from first commit to the last commit?

done
